kernel/src/circular_queue_kern.c

```diff
@@ -0,0 +1,277 @@
+
+  lck_spin_lock(queue->lck);
+  while (wait_result == THREAD_AWAKENED && queue->max_read == queue->read) {
+    wait_result = lck_spin_sleep(queue->lck, LCK_SLEEP_DEFAULT,
+                                 &queue->max_read, THREAD_ABORTSAFE);
+  }
+  size_t offset = queue->max_read - queue->buffer;
+  
+  lck_spin_unlock(queue->lck);
+  return offset;
+}
+
+int osquery_cqueue_dropped_data(osquery_cqueue_t *queue) {
+  int drops;
+  if (!queue->initialized) {
+    return -1;
+  }
+
+  lck_spin_lock(queue->lck);
+  drops = queue->drops;
+  queue->drops = 0;
+  lck_spin_unlock(queue->lck);
+
+  return drops;
+}
+
+void *osquery_cqueue_reserve(osquery_cqueue_t *queue, osquery_event_t event) {
+  if (!queue->initialized) {
+    return NULL;
+  }
+  
+  void *ret = NULL;
+  data_header_t *header = NULL;
+
+  size_t size = osquery_sizeof_event(event) + sizeof(data_header_t);
+
+  lck_spin_lock(queue->lck);
+  
+  // We do not want the write pointer to ever equal the read pointer unless
+  // everything is empty.  Otherwise we need to track the empty states for the
+  // buffer.
+  if (get_distance(queue, queue->write, queue->read, 1) > size) {
+    if (get_distance(queue, queue->write,
+                     queue->buffer + queue->size, 0) >= size) {
+      // We can fit the allocation by advancing the write pointer.
+      header = (data_header_t *)queue->write;
+      queue->write = (uint8_t *)advance_pointer(queue, queue->write, size);
+    } else if (get_distance(queue, queue->buffer, queue->read, 0) > size) {
+      // We can fit the allocation by wrapping the write pointer.
+      if (get_distance(queue, queue->write, queue->buffer + queue->size, 0)
+          >= sizeof(data_header_t *)) {
+        // Signal a Null event ie. jump to beginning of buf.
+        header = (data_header_t *)queue->write;
+        header->event = END_OF_BUFFER_EVENT;
+      }
+      header = (data_header_t *)queue->buffer;
+      queue->write = (uint8_t *)advance_pointer(queue, queue->buffer, size);
+    }
+  }
+
+  if (header) {
+    header->event = event;
+    header->finished = 0;
+
+    // Give them the pointer to the space not the header.
+    ret = (void *)(header + 1);
+  } else {
+    if (queue->drops >= 0) {
+      queue->drops += 1;
+    }
+    ret = NULL;
+  }
+
+  lck_spin_unlock(queue->lck);
+
+  return ret;
+}
+
+/** @brief Turn blocks that have been commited into readable space for user
+ *  level process.
+ *
+ *  REQUIRES the lock.
+ *
+ *  @param queue The queue to create readable space in.
+ *  @return Void.
+ */
+static inline void coalesce_readable(osquery_cqueue_t *queue) {
+  data_header_t *header = (data_header_t *)queue->max_read;
+  osquery_between_t b;
+
+  while (OSQUERY_BETWEEN & (b = is_between(queue, header, queue->max_read,
+                                       queue->write, sizeof(data_header_t)))) {
+    if (b & OSQUERY_NOT_IN_BUFFER || header->event == END_OF_BUFFER_EVENT) {
+      queue->max_read = queue->buffer;
```

Comment:
As per my comment above, this would be the case I believe given: `b & OSQUERY_NOT_IN_BUFFER`